src/libraries/Microsoft.PowerFx.Core/Types/DefinedTypeDependencyGraph.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.IR;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Syntax.Visitors;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Types
+{
+    internal class DefinedTypeDependencyGraph
+    {
+        private readonly IEnumerable<DefinedType> _definedTypes;
+        private readonly ReadOnlySymbolTable _globalTypes;
+
+        // stores unresolvedType -> its dependencies
+        private readonly Dictionary<DefinedType, HashSet<string>> _typeWithDependency;
+
+        // stores unresolvedType -> its dependendents
+        private readonly Dictionary<string, HashSet<DefinedType>> _invertedDependency;
+
+        // queue with types ready to be resolved
+        private readonly Queue<DefinedType> _tsQueue;
+
+        private readonly SymbolTable _definedTypeSymbolTable;
+
+        // TODO: Add more future type names
+        private static readonly ISet<string> _restrictedTypeNames = new HashSet<string> { "Record", "Table", "Currency" };
+
+        internal Dictionary<DefinedType, HashSet<string>> UnresolvedTypes => _typeWithDependency;
+
+        internal INameResolver DefinedTypeSymbols => ReadOnlySymbolTable.Compose(_definedTypeSymbolTable, _globalTypes);
+
+        public DefinedTypeDependencyGraph(IEnumerable<DefinedType> definedTypes, INameResolver globalSymbols) 
+        {
+            _definedTypes = definedTypes;
+            _globalTypes = ReadOnlySymbolTable.NewDefaultTypes(globalSymbols?.DefinedTypes);
+
+            _typeWithDependency = new Dictionary<DefinedType, HashSet<string>>();
+            _invertedDependency = new Dictionary<string, HashSet<DefinedType>>();
+            _tsQueue = new Queue<DefinedType>();
+
+            _definedTypeSymbolTable = new SymbolTable();
+
+            Build();
+        }
+
+        // Build type dependency graph to perform topological sort and resolve types
+        private void Build()
+        {
```

Comment:
Is the intent to rebuild this whole graph every time we re-parse App.Formulas? What conditions can we use to either ensure we only rebuild it when needed or can incrementally update it?

Reply:
Wouldn't NF optimization that we added help here? We only process UDFs which have changed. Need to be validated though.